cpp/src/arrow/CMakeLists.txt

```diff
@@ -301,13 +301,24 @@ arrow_add_pkg_config("arrow")
 #
 
 add_arrow_test(allocator-test)
-add_arrow_test(array-test
-               SOURCES
-               array-test.cc
-               array-binary-test.cc
-               array-dict-test.cc
-               array-list-test.cc
-               array-struct-test.cc)
+
+if(WIN32)
+  add_arrow_test(array-test
+                 SOURCES
+                 array-test.cc
+                 array-binary-test.cc
+                 array-list-test.cc
+                 array-struct-test.cc)
+else()
+  add_arrow_test(array-test
+                 SOURCES
+                 array-test.cc
+                 array-binary-test.cc
+                 array-dict-test.cc
+                 array-list-test.cc
+                 array-struct-test.cc)
+endif()
+
```

Comment:
Why is it needed?
Could you show error details?

Reply:
I don't know, I wasn't able to reproduce this one on my local MsysGW 32-bit environment, but let me disable it and hopefully your new test command will output something that can point us in the right direction.

Reply:
Alright, so from AppVeyor, looks like this is causing:

```
/arrow-array-test.dir/objects.a(array-dict-test.cc.obj):array-dict-test.cc:(.text+0xb9a2): undefined reference to `arrow::DictionaryBuilder<arrow::FixedSizeBinaryType>::DictionaryBuilder(std::shared_ptr<arrow::DataType> const&, arrow::MemoryPool*)'
CMakeFiles/arrow-array-test.dir/objects.a(array-dict-test.cc.obj):array-dict-test.cc:(.text+0xcb8a): undefined reference to `arrow::DictionaryBuilder<arrow::FixedSizeBinaryType>::DictionaryBuilder(std::shared_ptr<arrow::DataType> const&, arrow::MemoryPool*)'
CMakeFiles/arrow-array-test.dir/objects.a(array-dict-test.cc.obj):array-dict-test.cc:(.text+0xeef8): undefined reference to `arrow::DictionaryBuilder<arrow::FixedSizeBinaryType>::DictionaryBuilder(std::shared_ptr<arrow::DataType> const&, arrow::MemoryPool*)'
CMakeFiles/arrow-array-test.dir/objects.a(array-dict-test.cc.obj):array-dict-test.cc:(.text+0x10240): undefined reference to `arrow::DictionaryBuilder<arrow::FixedSizeBinaryType>::DictionaryBuilder(std::shared_ptr<arrow::DataType> const&, arrow::MemoryPool*)'
CMakeFiles/arrow-array-test.dir/objects.a(array-dict-test.cc.obj):array-dict-test.cc:(.text+0x104fc): undefined reference to `arrow::DictionaryBuilder<arrow::FixedSizeBinaryType>::AppendArray(arrow::Array const&)'
CMakeFiles/arrow-array-test.dir/objects.a(array-dict-test.cc.obj):array-dict-test.cc:(.text+0x108ef): undefined reference to `arrow

Reply:
Let me re-skip this test to have them back at passing... My suggestion would be to merge with a couple tests skipped and then open JIRA issues to investigate further. Otherwise, the longer it take us to merge this PR, the more likely something else will break Windows.

Reply:
Thanks.
We need to stop using template and `ARROW_EXPORT` for `arrow::DictionaryBuilder` like #3509.

Reply:
Could you open an issue on JIRA to work this on another pull request?
`if(WIN32)` is too general. It matches build with MSVC.
`if(WIN32 AND "${COMPILER_FAMIL}" STREQUAL "gcc")` is better.